src/feedbinapi.cpp

```diff
@@ -0,0 +1,321 @@
+		// requests, since it doesn't "guess" the correct auth type on the first
+		// try.
+		cfg.set_configvalue("http-auth-method", "basic");
+	}
+}
+
+bool FeedbinApi::authenticate()
+{
+	// error check handled in Controller
+	const Credentials creds = get_credentials("feedbin", "");
+	auth_info = strprintf::fmt("%s:%s", creds.user, creds.pass);
+
+	CurlHandle handle;
+	long response_code = 0;
+	run_op(FEEDBIN_AUTHENTICATION_PATH, json(), handle);
+	curl_easy_getinfo(handle.ptr(), CURLINFO_RESPONSE_CODE, &response_code);
+
+	return response_code == 200;
+}
+
+TaggedFeedUrl FeedbinApi::feed_from_json(const json& jfeed,
+	const std::vector<std::string>& addtags)
+{
+	const int feed_id = jfeed["feed_id"];
+	const std::string feed_title = jfeed["title"];
+	const std::string feed_url = jfeed["feed_url"];
+
+	std::vector<std::string> tags;
+	// automatically tag by feedtitle
+	tags.push_back(std::string("~") + feed_title);
+
+	// add additional tags
+	tags.insert(tags.end(), addtags.cbegin(), addtags.cend());
+
+	auto url = strprintf::fmt("%s#%d", feed_url, feed_id);
+	return TaggedFeedUrl(url, tags);
+}
+
+std::vector<TaggedFeedUrl> FeedbinApi::get_subscribed_urls()
+{
+	std::vector<TaggedFeedUrl> feeds;
+
+	const json taggings = run_op(FEEDBIN_TAGGINGS_PATH, json());
+	if (taggings.is_null()) {
+		LOG(Level::ERROR, "FeedbinApi::get_subscribed_urls: Failed to "
+			"retrieve taggings");
+		return feeds;
+	}
+
+	const json feed_list = run_op(FEEDBIN_SUBSCRIPTIONS_PATH, json());
+	if (feed_list.is_null()) {
+		LOG(Level::ERROR, "FeedbinApi::get_subscribed_urls: Failed to "
+			"retrieve feedlist");
+		return feeds;
+	}
+
+	for (const json& feed : feed_list) {
+		const int feed_id = feed["feed_id"];
+
+		std::vector<std::string> tags;
+		for (const auto& tagging : taggings) {
+			if (tagging["feed_id"] != feed_id) {
+				continue;
+			}
+
+			const std::string tag_name = tagging["name"];
+			tags.push_back(tagging["name"]);
+		}
+
+		feeds.push_back(feed_from_json(feed, tags));
+	}
+
+	return feeds;
+}
+
+bool FeedbinApi::mark_entries_read(const std::vector<std::string>& ids,
+	bool read)
+{
+	CurlHandle handle;
+	long response_code = 0;
+	HTTPMethod method = read ? HTTPMethod::DELETE : HTTPMethod::POST;
+	json body;
+	body["unread_entries"] = ids;
+	curl_slist* headers = NULL;
+	headers = curl_slist_append(headers, "Content-Type: application/json");
+	curl_easy_setopt(handle.ptr(), CURLOPT_HTTPHEADER, headers);
+	run_op(FEEDBIN_UNREAD_ENTRIES_PATH, body, handle, method);
+	curl_easy_getinfo(handle.ptr(), CURLINFO_RESPONSE_CODE, &response_code);
+
+	return response_code == 200;
+}
+
+bool FeedbinApi::mark_all_read(const std::string& combined_feed_url)
+{
+	std::string feed_id;
+	const std::string::size_type pound = combined_feed_url.find_first_of('#');
+	if (pound != std::string::npos) {
+		feed_id = combined_feed_url.substr(pound + 1);
+	} else {
+		LOG(Level::ERROR, "FeedbinApi::mark_all_read: Failed to "
+			"get feed ID from URL.");
```

Comment:
It'd be helpful to log the URL as well, so we can figure out why we couldn't get an ID from it.